### DISCUSSION TITLE
Whether system-config-printer could match the driver through the product fields in the PPD file ?

### DISCUSSION THREAD

**Discussion by PierceDong** [OpenPrinting/system-config-printer#370] [Q&A]
Whether system-config-printer could match the driver through the product fields in the PPD file ? 
For now, there is a printer Toshiba e-STUDIO 2520AC and the product fields in the PPD file already contains the info of this printer, but system-config-printer didn't match the driver automatically, from the code below, the driver of this printer should be matched successfully, right ?

aliases = {} # Generic model name: set(specific model names)
        for ppdname, ppddict in self.ppds.items ():
            # One entry for ppd-make-and-model
            ppd_make_and_model = _singleton (ppddict['ppd-make-and-model'])
            ppd_mm_split = ppdMakeModelSplit (ppd_make_and_model)
            ppd_makes_and_models = set([ppd_mm_split])

            ****# The ppd-product IPP attribute contains values from each
            # Product PPD attribute as well as the value from the
            # ModelName attribute if present.  The Product attribute
            # values are surrounded by parentheses; the ModelName
            # attribute value is not.****

            **# Add another entry for each ppd-product that came from a
            # Product attribute in the PPD file.**
            ppd_products = ppddict.get ('ppd-product', [])
            if not isinstance (ppd_products, list):
                ppd_products = [ppd_products]
            ppd_products = set ([x for x in ppd_products if x.startswith ("(")])
            if ppd_products:
                # If there is only one ppd-product value it is
                # unlikely to be useful.
                if len (ppd_products) == 1:
                    ppd_products = set()

from the comments above. the code will process the product fields in the PPD file, right ?
the Product fields in the PPD file looks like this:

*Product: "(TOSHIBA e-STUDIO2520AC)"
*Product: "(TOSHIBA e-STUDIO5525AC)"
*Product: "(TOSHIBA e-STUDIO4525AC)"
*Product: "(TOSHIBA e-STUDIO3525AC)"
*Product: "(TOSHIBA e-STUDIO3025AC)"
*Product: "(TOSHIBA e-STUDIO2525AC)"
*Product: "(TOSHIBA e-STUDIO2025AC)"
*Product: "(TOSHIBA e-STUDIO2520AC)"

**Comment by PierceDong** [comment]
Really appreciate your help.

**Reply by SinghCod3r** [reply]
Hi!, @PierceDong 

Yes, system-config-printer does process the `*Product: `entries. It strips the parentheses, adds the manufacturer if necessary, converts each product into a make/model pair with `ppdMakeModelSplit()` , and includes them as aliases during PPD initialization. However, automatic driver selection is performed later in `getPPDNamesFromDeviceID()` , where the detected **IEEE-1284 Device ID** is matched against these normalized make/model entries. If the Toshiba driver is still not selected, the next step would be to compare the printer's actual Device ID **(MFG/MDL)** with the normalized PPD entries to see where the comparison fails.
